File: TelerikUIExamplesInSwift/TelerikUIExamplesInSwift/MultipleAxes.swift
//
//  MultipleAxes.swift
//  Swift Examples
//
//  Copyright (c) 2014 Telerik. All rights reserved.
//

import Foundation

class MultipleAxes:ExampleViewController {
    
    let chart = TKChart()
    
    override func viewDidLoad() {
        super.viewDidLoad()
    
        chart.frame = self.exampleBoundsWithInset
        chart.autoresizingMask = UIViewAutoresizing.FlexibleWidth | UIViewAutoresizing.FlexibleHeight
        self.view.addSubview(chart)
    
        let gdpInPoundsYAxis = TKChartNumericAxis(minimum: 1050, andMaximum: 1400)
        gdpInPoundsYAxis.majorTickInterval = 50
        gdpInPoundsYAxis.position = TKChartAxisPosition.Left
        gdpInPoundsYAxis.style.labelStyle.textAlignment = TKChartAxisLabelAlignment.Left
        gdpInPoundsYAxis.style.majorTickStyle.ticksHidden = false
        gdpInPoundsYAxis.style.lineHidden = false
        chart.addAxis(gdpInPoundsYAxis)
        chart.yAxis = gdpInPoundsYAxis
        
        let periodXAxis = TKChartDateTimeAxis()
        periodXAxis.majorTickIntervalUnit = TKChartDateTimeAxisIntervalUnit.Years
        periodXAxis.majorTickInterval = 1
        periodXAxis.position = TKChartAxisPosition.Bottom
        periodXAxis.setPlotMode(TKChartAxisPlotMode.BetweenTicks)
        chart.addAxis(periodXAxis)
        
        let gdpInvestmentYAxis = TKChartNumericAxis(minimum: 0, andMaximum: 20)
        gdpInvestmentYAxis.majorTickInterval = 5
        gdpInvestmentYAxis.position = TKChartAxisPosition.Right
        gdpInvestmentYAxis.style.labelStyle.textAlignment = TKChartAxisLabelAlignment.Left
        gdpInvestmentYAxis.style.majorTickStyle.ticksHidden = false
        gdpInvestmentYAxis.style.lineHidden = false
        chart.addAxis(gdpInvestmentYAxis)
        
        let gdpGrowthUpAnnualChangeYAxis = TKChartNumericAxis(minimum:-6, andMaximum:4)
        gdpGrowthUpAnnualChangeYAxis.majorTickInterval = 1
        gdpGrowthUpAnnualChangeYAxis.position = TKChartAxisPosition.Right
        gdpGrowthUpAnnualChangeYAxis.style.labelStyle.textAlignment = TKChartAxisLabelAlignment.Left
        gdpGrowthUpAnnualChangeYAxis.style.majorTickStyle.ticksHidden = false
        gdpGrowthUpAnnualChangeYAxis.style.lineHidden = false
        chart.addAxis(gdpGrowthUpAnnualChangeYAxis)
        
        let grossNationalSavingsAnnualGrowthUpYAxis = TKChartNumericAxis(minimum:0, andMaximum:16)
        grossNationalSavingsAnnualGrowthUpYAxis.majorTickInterval = 2
        grossNationalSavingsAnnualGrowthUpYAxis.position = TKChartAxisPosition.Right
        grossNationalSavingsAnnualGrowthUpYAxis.style.labelStyle.textAlignment = TKChartAxisLabelAlignment.Left
        grossNationalSavingsAnnualGrowthUpYAxis.style.majorTickStyle.ticksHidden = false
        grossNationalSavingsAnnualGrowthUpYAxis.style.lineHidden = false
        chart.addAxis(grossNationalSavingsAnnualGrowthUpYAxis)
        
        let date2001:NSDate = MultipleAxes.dateWithYear(2001, month: 12, day: 31)
        let date2002:NSDate = MultipleAxes.dateWithYear(2002, month: 12, day: 31)
        let date2003:NSDate = MultipleAxes.dateWithYear(2003, month: 12, day: 31)
        let date2004:NSDate = MultipleAxes.dateWithYear(2004, month: 12, day: 31)
        let date2005:NSDate = MultipleAxes.dateWithYear(2005, month: 12, day: 31)
        
        let gdpInPounds = [TKChartDataPoint(x:date2001, y:1200),
            TKChartDataPoint(x:date2002, y:1200),
            TKChartDataPoint(x:date2003, y:1225),
            TKChartDataPoint(x:date2004, y:1300),
            TKChartDataPoint(x:date2005, y:1350)]
        
        let gdpInPoundsSeries = TKChartColumnSeries(items: gdpInPounds)
        gdpInPoundsSeries.xAxis = periodXAxis
        gdpInPoundsSeries.yAxis = gdpInPoundsYAxis
        gdpInPoundsSeries.selectionMode = TKChartSeriesSelectionMode.Series
        chart.addSeries(gdpInPoundsSeries)
        
        let gdpGrowthUpAnnual = [TKChartDataPoint(x:date2001, y:4),
            TKChartDataPoint(x:date2002, y:3),
            TKChartDataPoint(x:date2003, y:2),
            TKChartDataPoint(x:date2004, y:-5),
            TKChartDataPoint(x:date2005, y:1)]
        
        let shapeSize = CGFloat(UIDevice.currentDevice().userInterfaceIdiom == .Phone ? 7.0 : 17.0);
        let gdpGrowthUpSeries = TKChartLineSeries(items:gdpGrowthUpAnnual)
        gdpGrowthUpSeries.style.pointShape = TKPredefinedShape(type:TKShapeType.Circle, andSize: CGSizeMake(shapeSize, shapeSize))
        gdpGrowthUpSeries.xAxis = periodXAxis
        gdpGrowthUpSeries.yAxis = gdpGrowthUpAnnualChangeYAxis
        gdpGrowthUpSeries.selectionMode = TKChartSeriesSelectionMode.Series
        gdpGrowthUpSeries.style.shapeMode = TKChartSeriesStyleShapeMode.AlwaysShow
        chart.addSeries(gdpGrowthUpSeries)
        
        let grossAnualSavings = [TKChartDataPoint(x:date2001, y:14),
            TKChartDataPoint(x:date2002, y:8),
            TKChartDataPoint(x:date2003, y:12),
            TKChartDataPoint(x:date2004, y:11),
            TKChartDataPoint(x:date2005, y:16)]
        
        let grossAnualSavingsSeries = TKChartLineSeries(items: grossAnualSavings)
        grossAnualSavingsSeries.style.pointShape = TKPredefinedShape(type:TKShapeType.Circle, andSize:CGSizeMake(shapeSize, shapeSize))
        grossAnualSavingsSeries.xAxis = periodXAxis
        grossAnualSavingsSeries.yAxis = grossNationalSavingsAnnualGrowthUpYAxis
        grossAnualSavingsSeries.selectionMode = TKChartSeriesSelectionMode.Series
        grossAnualSavingsSeries.style.shapeMode = TKChartSeriesStyleShapeMode.AlwaysShow
        chart.addSeries(grossAnualSavingsSeries)
        
        let gdpInvestment = [TKChartDataPoint(x:date2001, y:15),
                            TKChartDataPoint(x:date2002, y:13),
                            TKChartDataPoint(x:date2003, y:16),
                            TKChartDataPoint(x:date2004, y:19),
                            TKChartDataPoint(x:date2005, y:15)]
        
        let gdpInvestmentSeries = TKChartLineSeries(items:gdpInvestment)
        gdpInvestmentSeries.style.pointShape = TKPredefinedShape(type:TKShapeType.Circle, andSize:CGSizeMake(shapeSize, shapeSize))
        gdpInvestmentSeries.xAxis = periodXAxis
        gdpInvestmentSeries.yAxis = gdpInPoundsYAxis
        gdpInvestmentSeries.selectionMode = TKChartSeriesSelectionMode.Series
        gdpInvestmentSeries.style.shapeMode = TKChartSeriesStyleShapeMode.AlwaysShow
        chart.addSeries(gdpInvestmentSeries)
        
        MultipleAxes.setStyle(gdpInPoundsSeries)
        MultipleAxes.setStyle(grossAnualSavingsSeries)
        MultipleAxes.setStyle(gdpGrowthUpSeries)
        MultipleAxes.setStyle(gdpInvestmentSeries)
        
        chart.reloadData()
    }
    
    class func dateWithYear(year: Int, month: Int, day: Int) -> NSDate {
        var calendar = NSCalendar(calendarIdentifier: NSGregorianCalendar)!
        var components = NSDateComponents()
        components.year = year
        components.month = month
        components.day = day
        return calendar.dateFromComponents(components)!
    }
    
    class func setStyle(series:TKChartSeries) {
        let item = series.style.palette.paletteItemAtIndex(series.index)
        
        if series.isKindOfClass(TKChartColumnSeries) && item.drawables.count > 1 {
            let drawable = item.drawables[2] as! TKDrawing
            
            if drawable.isKindOfClass(TKStroke){
                var stroke = drawable as! TKStroke
                series.yAxis.style.lineStroke = TKStroke(fill: stroke.fill)
            }
            else {
                series.yAxis.style.lineStroke = TKStroke(fill: item.stroke.fill)
            }
        }
        else {
            series.yAxis.style.lineStroke = TKStroke(fill: item.stroke.fill)
        }
        
        series.yAxis.style.majorTickStyle.ticksFill = series.yAxis.style.lineStroke.fill
        series.yAxis.style.majorTickStyle.maxTickClippingMode = TKChartAxisClippingMode.Visible
        series.yAxis.style.majorTickStyle.minTickClippingMode = TKChartAxisClippingMode.Visible
        
        if series.yAxis.style.majorTickStyle.ticksFill.isKindOfClass(TKSolidFill) {
            let solidFill = series.yAxis.style.majorTickStyle.ticksFill as! TKSolidFill
            series.yAxis.style.labelStyle.textColor = solidFill.color
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
}

File: Examples/Examples/Docs/AxisTypes/MultipleAxes.cs
using System;
using System.Collections.Generic;
using System.Drawing;

using Foundation;
using UIKit;

using TelerikUI;

namespace Examples
{
	[Register("MultipleAxes")]
	public class MultipleAxes: XamarinExampleViewController
	{
		public override void ViewDidLoad ()
		{
			base.ViewDidLoad ();

			TKChart chart = new TKChart (this.View.Bounds);
			chart.AutoresizingMask = UIViewAutoresizing.FlexibleWidth | UIViewAutoresizing.FlexibleHeight;
			this.View.AddSubview (chart);

			TKChartNumericAxis gdpInPoundsYAxis = new TKChartNumericAxis (new NSNumber(1050), new NSNumber(1400));
			gdpInPoundsYAxis.MajorTickInterval = 50;
			gdpInPoundsYAxis.Position = TKChartAxisPosition.Left;
			gdpInPoundsYAxis.Style.LabelStyle.TextAlignment = TKChartAxisLabelAlignment.Left;
			gdpInPoundsYAxis.Style.MajorTickStyle.TicksHidden = false;
			gdpInPoundsYAxis.Style.LineHidden = false;
			chart.AddAxis (gdpInPoundsYAxis);
			chart.YAxis = gdpInPoundsYAxis;

			// >> chart-axis-position-cs
			TKChartDateTimeAxis periodXAxis = new TKChartDateTimeAxis ();
			periodXAxis.MajorTickIntervalUnit = TKChartDateTimeAxisIntervalUnit.Years;
			periodXAxis.MinorTickIntervalUnit = TKChartDateTimeAxisIntervalUnit.Years;
			periodXAxis.MajorTickInterval = 1;
			periodXAxis.Position = TKChartAxisPosition.Bottom;
			periodXAxis.PlotMode = TKChartAxisPlotMode.BetweenTicks;
			chart.AddAxis (periodXAxis);

			TKChartNumericAxis gdpInvestmentYAxis = new TKChartNumericAxis (new NSNumber(0), new NSNumber(20));
			gdpInvestmentYAxis.MajorTickInterval = 5;
			gdpInvestmentYAxis.Position = TKChartAxisPosition.Right;
			gdpInvestmentYAxis.Style.LabelStyle.TextAlignment = TKChartAxisLabelAlignment.Left;
			gdpInvestmentYAxis.Style.MajorTickStyle.TicksHidden = false;
			gdpInvestmentYAxis.Style.LineHidden = false;
			chart.AddAxis (gdpInvestmentYAxis);
			// << chart-axis-position-cs

			TKChartNumericAxis gdpGrowthUpAnnualChangeYAxis = new TKChartNumericAxis (new NSNumber(-6), new NSNumber(4));
			gdpGrowthUpAnnualChangeYAxis.MajorTickInterval = 1;
			gdpGrowthUpAnnualChangeYAxis.Position = TKChartAxisPosition.Right;
			gdpGrowthUpAnnualChangeYAxis.Style.LabelStyle.TextAlignment = TKChartAxisLabelAlignment.Left;
			gdpGrowthUpAnnualChangeYAxis.Style.MajorTickStyle.TicksHidden = false;
			gdpGrowthUpAnnualChangeYAxis.Style.LineHidden = false;
			chart.AddAxis (gdpGrowthUpAnnualChangeYAxis);

			TKChartNumericAxis grossNationalSavingsAnnualGrowthUpYAxis = new TKChartNumericAxis (new NSNumber(0), new NSNumber(16));
			grossNationalSavingsAnnualGrowthUpYAxis.MajorTickInterval = 2;
			grossNationalSavingsAnnualGrowthUpYAxis.Position = TKChartAxisPosition.Right;
			grossNationalSavingsAnnualGrowthUpYAxis.Style.LabelStyle.TextAlignment = TKChartAxisLabelAlignment.Left;
			grossNationalSavingsAnnualGrowthUpYAxis.Style.MajorTickStyle.TicksHidden = false;
			grossNationalSavingsAnnualGrowthUpYAxis.Style.LineHidden = false;
			chart.AddAxis (grossNationalSavingsAnnualGrowthUpYAxis);

			NSDate date2001 = DateWithYear (2001, 12, 31);
			NSDate date2002 = DateWithYear (2002, 12, 31);
			NSDate date2003 = DateWithYear (2003, 12, 31);
			NSDate date2004 = DateWithYear (2004, 12, 31);
			NSDate date2005 = DateWithYear (2005, 12, 31);

			TKChartDataPoint[] gdpInPounds = new TKChartDataPoint[] {
				new TKChartDataPoint(date2001, new NSNumber(1200)),
				new TKChartDataPoint(date2002, new NSNumber(1200)),
				new TKChartDataPoint(date2003, new NSNumber(1225)),
				new TKChartDataPoint(date2004, new NSNumber(1300)),
				new TKChartDataPoint(date2005, new NSNumber(1350))
			};

			TKChartColumnSeries gdpInPoundsSeries = new TKChartColumnSeries  (gdpInPounds);
			gdpInPoundsSeries.XAxis = periodXAxis;
			gdpInPoundsSeries.YAxis = gdpInPoundsYAxis;
			gdpInPoundsSeries.SelectionMode = TKChartSeriesSelectionMode.Series;
			chart.AddSeries (gdpInPoundsSeries);

			TKChartDataPoint[] gdpGrowthUpAnnual = new TKChartDataPoint[] {
				new TKChartDataPoint (date2001, new NSNumber (4)),
				new TKChartDataPoint (date2002, new NSNumber (3)),
				new TKChartDataPoint (date2003, new NSNumber (2)),
				new TKChartDataPoint (date2004, new NSNumber (-5)),
				new TKChartDataPoint (date2005, new NSNumber (1))
			};

			float shapeSize = UIDevice.CurrentDevice.UserInterfaceIdiom == UIUserInterfaceIdiom.Phone ? 7 : 17;

			TKChartLineSeries gdpGrowthUpSeries = new TKChartLineSeries (gdpGrowthUpAnnual);
			gdpGrowthUpSeries.Style.PointShape = new TKPredefinedShape (TKShapeType.Circle, new SizeF (shapeSize, shapeSize));
			gdpGrowthUpSeries.XAxis = periodXAxis;
			gdpGrowthUpSeries.YAxis = gdpGrowthUpAnnualChangeYAxis;
			gdpGrowthUpSeries.SelectionMode = TKChartSeriesSelectionMode.DataPoint;
			gdpGrowthUpSeries.Style.ShapeMode = TKChartSeriesStyleShapeMode.AlwaysShow;
			chart.AddSeries (gdpGrowthUpSeries);

			TKChartDataPoint[] grossAnualSavings = new TKChartDataPoint[] {
				new TKChartDataPoint (date2001, new NSNumber (14)),
				new TKChartDataPoint (date2002, new NSNumber (8)),
				new TKChartDataPoint (date2003, new NSNumber (12)),
				new TKChartDataPoint (date2004, new NSNumber (11)),
				new TKChartDataPoint (date2005, new NSNumber (16))
			};

			TKChartLineSeries grossAnualSavingsSeries = new TKChartLineSeries (grossAnualSavings);
			grossAnualSavingsSeries.Style.PointShape = new TKPredefinedShape (TKShapeType.Circle, new SizeF(shapeSize, shapeSize));
			grossAnualSavingsSeries.XAxis = periodXAxis;
			grossAnualSavingsSeries.YAxis = grossNationalSavingsAnnualGrowthUpYAxis;
			grossAnualSavingsSeries.SelectionMode = TKChartSeriesSelectionMode.DataPoint;
			grossAnualSavingsSeries.Style.ShapeMode = TKChartSeriesStyleShapeMode.AlwaysShow;
			chart.AddSeries (grossAnualSavingsSeries);

			TKChartDataPoint[] gdpInvestment = new TKChartDataPoint[] {
				new TKChartDataPoint (date2001, new NSNumber (15)),
				new TKChartDataPoint (date2002, new NSNumber (13)),
				new TKChartDataPoint (date2003, new NSNumber (16)),
				new TKChartDataPoint (date2004, new NSNumber (19)),
				new TKChartDataPoint (date2005, new NSNumber (15))
			};

			TKChartLineSeries gdpInvestmentSeries = new TKChartLineSeries (gdpInvestment);
			gdpInvestmentSeries.Style.PointShape = new TKPredefinedShape (TKShapeType.Circle, new SizeF(shapeSize, shapeSize));
			gdpInvestmentSeries.XAxis = periodXAxis;
			gdpInvestmentSeries.YAxis = gdpInvestmentYAxis;
			gdpInvestmentSeries.SelectionMode = TKChartSeriesSelectionMode.DataPoint;
			gdpInvestmentSeries.Style.ShapeMode = TKChartSeriesStyleShapeMode.AlwaysShow;
			chart.AddSeries (gdpInvestmentSeries);

			this.setStyles (gdpInPoundsSeries);
			this.setStyles (grossAnualSavingsSeries);
			this.setStyles (gdpGrowthUpSeries);
			this.setStyles (gdpInvestmentSeries);

			chart.ReloadData ();
		}

		public void setStyles(TKChartSeries series)
		{
			TKChartPaletteItem item = series.Style.Palette.PaletteItemAtIndex (series.Index);
			if (series is TKChartColumnSeries && item.Drawables.Length > 1) {
				NSObject drawable = item.Drawables [2];

				if (drawable is TKStroke) {
					TKStroke stroke = drawable as TKStroke;
					series.YAxis.Style.LineStroke = new TKStroke (stroke.Fill);
				} 
				else {
					series.YAxis.Style.LineStroke = new TKStroke (item.Stroke.Fill);
				} 
			}
			else {
				series.YAxis.Style.LineStroke = new TKStroke(item.Stroke.Fill);
			}
			series.YAxis.Style.MajorTickStyle.TicksFill = series.YAxis.Style.LineStroke.Fill;

			// >> chart-tick-style-clipping-cs
			series.YAxis.Style.MajorTickStyle.MaxTickClippingMode = TKChartAxisClippingMode.Visible;
			series.YAxis.Style.MajorTickStyle.MinTickClippingMode = TKChartAxisClippingMode.Visible;
			// << chart-tick-style-clipping-cs

			if (series.YAxis.Style.MajorTickStyle.TicksFill is TKSolidFill) {
				TKSolidFill solidFill = series.YAxis.Style.MajorTickStyle.TicksFill as TKSolidFill;
				series.YAxis.Style.LabelStyle.TextColor = solidFill.Color;
			}

		}

		public NSDate DateWithYear(int year, int month, int day)
		{
			NSCalendar calendar = new NSCalendar (NSCalendarType.Gregorian);
			NSDateComponents components = new NSDateComponents ();
			components.Year = year;
			components.Month = month;
			components.Day = day;
			return calendar.DateFromComponents (components);
		}
	}
}

